src/theory/sets/theory_sets_type_enumerator.h

```diff
@@ -2,185 +2,66 @@
-
-    return n;
-  }
-
-  SetEnumerator& operator++() override
-  {
-    Trace("set-type-enum") << "operator++ called, **this = " << **this << std::endl;
-
-    if (d_finished) {
-      Trace("set-type-enum") << "operator++ finished!" << std::endl;
-      return *this;
-    }
-
-    // increment by one, at the same time deleting all elements that
-    // cannot be incremented any further (note: we are keeping a set
-    // -- no repetitions -- thus some trickery to know what to pop and
-    // what not to.)
-    if(d_index > 0) {
-      Assert(d_index == d_constituentVec.size());
-
-      Node last_pre_increment;
-      last_pre_increment = *(*d_constituentVec.back());
-
-      ++(*d_constituentVec.back());
-
-      if (d_constituentVec.back()->isFinished()) {
-        delete d_constituentVec.back();
-        d_constituentVec.pop_back();
-
-        while(!d_constituentVec.empty()) {
-          Node cur_pre_increment = *(*d_constituentVec.back());
-          ++(*d_constituentVec.back());
-          Node cur_post_increment = *(*d_constituentVec.back());
-          if(last_pre_increment == cur_post_increment) {
-            delete d_constituentVec.back();
-            d_constituentVec.pop_back();
-            last_pre_increment = cur_pre_increment;
-          } else {
-            break;
-          }
-        }
-      }
-    }
-
-    if (d_constituentVec.empty()) {
-      ++d_index;
-      d_constituentVec.push_back(new TypeEnumerator(d_constituentType, d_tep));
-    }
-
-    while (d_constituentVec.size() < d_index) {
-      TypeEnumerator* newEnumerator =
-          new TypeEnumerator(*d_constituentVec.back());
-      ++(*newEnumerator);
-      if (newEnumerator->isFinished()) {
-        Trace("set-type-enum") << "operator++ finished!" << std::endl;
-        delete newEnumerator;
-        d_finished = true;
-        return *this;
-      }
-      d_constituentVec.push_back(newEnumerator);
-    }
-
-    Trace("set-type-enum") << "operator++ returning, **this = " << **this << std::endl;
-    return *this;
-  }
-
-  bool isFinished() override
-  {
-    Trace("set-type-enum") << "isFinished returning: " << d_finished << std::endl;
-    return d_finished;
-  }
-
-};/* class SetEnumerator */
-
-}/* CVC4::theory::sets namespace */
-}/* CVC4::theory namespace */
-}/* CVC4 namespace */
+  SetEnumerator(TypeNode type, TypeEnumeratorProperties* tep = nullptr);
+  SetEnumerator(const SetEnumerator& enumerator);
+  ~SetEnumerator();
+
+  Node operator*() override;
+
+  /**
+   * This operator iterates over the power set of the element type
+   * following the order of a bitvector counter.
+   * Example: iterating over a set of bitvecotors of length 2 will return the
+   * following sequence consisting of 16 sets:
+   * {}, {00}, {01}, {00, 01}, {10}, {00, 10}, {01, 10}, {00, 01, 10}, ...,
+   * {00, 01, 10, 11}
+   */
+  SetEnumerator& operator++() override;
+
+  bool isFinished() override;
+
+ private:
+  NodeManager* d_nodeManager;
```

Comment:
Can you add a one line comment for each of these?

Reply:
done